### ISSUE TITLE
Refactor magisk-modules-repo-util

### ISSUE THREAD

**Issue by ya0211** [ya0211/magisk-modules-repo-util#1]
###  Features

- [x] Add license field in JSON (for app)
- [x] Add update.json for each module and have its version control
- [x] A new config.json  
- [x] Log output and storage are controllable
- [x] Keep the `states` field (zipUrl and changelog), compatible with the old version of MRepo
- [x] Support sync model: `git`
- [ ] Support sync model: `rsync`

**Comment by IzzySoft** [comment]
> Variables of Repo will be done in a config file (for git)

Not only "for git", but for all variables a user might need to change (consider e.g. I don't want to use Github etc to host my repo, but simply use `rsync` to push the `modules/` and `json/` directories to a server I control; I'd need to define the remote host and directory, and that I want to use `rsync` instead of `git`).

I'd suggest to use a `config.json` or (as F-Droid does) a `config.yml` for this (and have the repo ship a `config.json.sample` a maintainer could use to get started). Taking a look at F-Droid's `config.yml` and the `modules.json`s of established Magisk repos and MRepo, I could imagine something like this (with the values adjusted of course – they do not reflect my real structures anyway):

```yaml
repo_url: https://apt.izzysoft.de/magisk
repo_name: IzzyOnDroid Magisk Repo
serverwebroot: apt.izzysoft.de:/var/www/magisk
sync_mode: rsync
```

or as JSON (as the other data structures like `modules.json` use this as well – so lets keep things simple and stick with that):

```json
{
  "repo_url": "https://apt.izzysoft.de/magisk",
  "repo_name": "IzzyOnDroid Magisk Repo",
  "serverwebroot": "apt.izzysoft.de:/var/www/magisk",
  "sync_mode": "rsync"
}
```

This would be an example for my constellation, using `rsync` to upload the files to my server. For Github, the `repo_url` would then be something like `https://raw.githubusercontent.com/{user}/{repo}/{branch}`, and `sync_mode: git`. For Gogs, Gitea & Forgejo, `https://{host}}/{user}/{repo}/raw/branch/{branch}`, for GitLab `https://{host}/{user}/{repo}/-/raw/{branch}` and so on. For `sync_mode: rsync` there could be a pendant to `sync.push()` taking care for `rsync`:

```python
    def rsync(self):
        msg = "timestamp: {0}".format(datetime.now())
        subprocess.run(['rsync', '-ae', 'ssh', '.', "{0}/{1}/".format(serverwebroot,"modules"], cwd=self.modules_folder.as_posix())
        subprocess.run(['rsync', '-ae', 'ssh', '.', "{0}/{1}/".format(serverwebroot,"json"], cwd=self.json_folder.as_posix())
```

`sync_mode` could also be `none`, in which case the structures would remain local – and the maintainers could use any means of syncing they prefer in a separate step (e.g. scripting).

PS: I've not tried that Python code for `def rsync()`, so please cross-check if you wish to copy it :wink: Oh, and the `config.json` would then be expected directly in `root_folder` I guess? So maybe this repo should be reorganized a little, moving all the `*.py` files back into a `util/` directory to reflect that – so later there could also be directories like `doc/` or `examples/`.

**Comment by ya0211** [comment]
> > Variables of Repo will be done in a config file (for git)
> 
> Not only "for git", but for all variables a user might need to change (consider e.g. I don't want to use Github etc to host my repo, but simply use `rsync` to push the `modules/` and `json/` directories to a server I control; I'd need to define the remote host and directory, and that I want to use `rsync` instead of `git`).
> 
> I'd suggest to use a `config.json` or (as F-Droid does) a `config.yml` for this (and have the repo ship a `config.json.sample` a maintainer could use to get started). Taking a look at F-Droid's `config.yml` and the `modules.json`s of established Magisk repos and MRepo, I could imagine something like this (with the values adjusted of course – they do not reflect my real structures anyway):
> 
> ```yaml
> repo_url: https://apt.izzysoft.de/magisk
> repo_name: IzzyOnDroid Magisk Repo
> serverwebroot: apt.izzysoft.de:/var/www/magisk
> sync_mode: rsync
> ```
> 
> or as JSON (as the other data structures like `modules.json` use this as well – so lets keep things simple and stick with that):
> 
> ```json
> {
>   "repo_url": "https://apt.izzysoft.de/magisk",
>   "repo_name": "IzzyOnDroid Magisk Repo",
>   "serverwebroot": "apt.izzysoft.de:/var/www/magisk",
>   "sync_mode": "rsync"
> }
> ```
> 
> This would be an example for my constellation, using `rsync` to upload the files to my server. For Github, the `repo_url` would then be something like `https://raw.githubusercontent.com/{user}/{repo}/{branch}`, and `sync_mode: git`. For Gogs, Gitea & Forgejo, `https://{host}}/{user}/{repo}/raw/branch/{branch}`, for GitLab `https://{host}/{user}/{repo}/-/raw/{branch}` and so on. For `sync_mode: rsync` there could be a pendant to `sync.push()` taking care for `rsync`:
> 
> ```python
>     def rsync(self):
>         msg = "timestamp: {0}".format(datetime.now())
>         subprocess.run(['rsync', '-ae', 'ssh', '.', "{0}/{1}/".format(self.serverwebroot,"modules"], cwd=self.modules_folder.as_posix())
>         subprocess.run(['rsync', '-ae', 'ssh', '.', "{0}/{1}/".format(self.serverwebroot,"json"], cwd=self.json_folder.as_posix())
> ```
> 
> `sync_mode` could also be `none`, in which case the structures would remain local – and the maintainers could use any means of syncing they prefer in a separate step (e.g. scripting).
> 
> PS: I've not tried that Python code for `def rsync()`, so please cross-check if you wish to copy it 😉 Oh, and the `config.json` would then be expected directly in `root_folder` I guess? So maybe this repo should be reorganized a little, moving all the `*.py` files back into a `util/` directory to reflect that – so later there could also be directories like `doc/` or `examples/`.

A new sync program has been submitted. 🥳

**Comment by IzzySoft** [comment]
Thanks! But the split with `BASE_URL` etc. makes it a bit complicated/confusing (especially as there's no description to it). Going by the example of your repo: If I'd only set `RAW_URL` and leave the other 3 empty, wouldn't that work as well? Like

```json
{
  "RAW_URL": "https://raw.githubusercontent.com/{user}/{repo}/{branch}"
}
```

and be done? (answer: unfortunately not, as `sync.py` crashes if not all fields are present; will send a fix as PR: #2). Besides, any reason the names must be upper-case here?

And considering other means of sync (like `rsync`), shouldn't the name rather be `REPO_URL`? Or is that now `BASE_URL`? And does your `"RAW_URL": "BASE_URL/MODULE_REPO/raw/REPO_BRANCH"` really work (its a string, so how does it tell which part are variables and which are literals?I could have a repo `github.com/izzy/MODULE_REPO`, or my repo could have a branch named `MODULE_REPO`. So should your code now use a simple search-and-replace on the string, that might break things. I'd discourage from that.

I first liked the separated `update.json` files per module – thinking those would be the only ones one needs to edit, and the `modules.json` generated based on their content. And as they now seem to support multiple versions, shouldn't the `config.json` have a setting of how many versions to keep? Or is it just the metadata which is kept, the ZIP file kept is the one in `hosts.json`, and it's the latter one we must edit to e.g. add a new module? Looks like that, as that's the only file having the `license` field. And the `zipUrl` and `changelog` are now entirely missing from `modules.json`, making it necessary to look up all `update.json` files separately if one e.g. wants to create a web front-end including the download URL. Would you consider adding those fields back to `modules.json`?

Speaking of license, just a hint: `GPL-3.0` is ambiguous. Depending on the declaration, it could be just GPLv3, or also include any later versions. Hence the prefixes `-only` and `-or-later` are used at F-Droid (and probably also in other places). `zygisk_lsposed` e.g. is `GPL-3.0-only` :wink:

One last thing concerns logging. 3 log files each run, isn't that overdoing things? Why not using a single log file, and leaving it to e.g. `logrotate` to clean up? Maybe cosmetical, as currently one could also just delete files "older than" – but still **3** logs per run?

**Comment by ya0211** [comment]
### config.json
- using yaml is not currently considered, [PyYaml](https://pyyaml.org/) is not a python built-in package and needs to be installed (although this is simple)
- there should be a new config.json
```json
{
  "sync_mode": "git",
  "repo_name": "Sanmer Magisk Repo",
  "repo_url": "https://raw.githubusercontent.com/ya0211/magisk-modules-repo/main/",
  "max_num_module": "3",
  "show_log": "false"
}
```
- `sync_mode`: git, rsync etc.
- `raw_url`:  rename to `repo_url` and it will be defined by the user
- `max_num_module`: the maximum number of old version modules to keep, currently this is a constant defined in repo.py

### hosts.json
- hosts.json and config.json(only one time) need to be defined by the user
- adding new modules needs to be done here
```json
[
    {
      "id": "zygisk_lsposed",
      "update_to": "https://lsposed.github.io/LSPosed/release/zygisk.json",
      "license": "GPL-3.0-only"
    }
]
```
- `name`: remove, it's defined by config.json

### zipUrl and changelog
now, when using these data, the program needs to send a request，for MRepo this is ok.
If we restore them to modules.json there are only two ways
- just like before only the latest
- merge update.json into modules.json


### other
- the problem of too many log files, I will merge them into one
- thanks for the answer about the license, I don't know the details very well 🧐

**Comment by IzzySoft** [comment]
> using yaml is not currently considered

Nope, sure not, full ack – as I wrote "lets keep things simple and stick with [JSON]". I had just looked up the fdroidserver config for inspiration, which uses YAML, and then rewrote it to JSON :wink:

> there should be a new config.json

:+1: Yes, that looks great! Then I'll revert my changes (the PR you closed; I first was disappointed/discouraged, but it seems we were working in parallel and you already had a more complete rewrite in the works).

> zipUrl and changelog

"just like before only the latest" would probably suffice: when generating a list of all modules, you probably want to link the latest version for download and that's it. When showing a "details page" it would just imply reading one additional file, which is fine.

On the other hand, considering a "remote client" (like MRepo) wants to display a details page, it would involve fetching all those files; here it would be better if just a single "index" (`modules.json`) would contain all details needed. I'm fine with both approaches; but I'd tend to having a "full index" in a single file. Not sure how the "established repos" deal with multiple versions, though (and our roadmap here includes a "compatible index" IIRC, which then of course could be another file.

> the problem of too many log files, I will merge them into one

Great! I also see the `show_log` option. Didn't check the code yet: is that just for the output on the terminal (to "silence" the output e.g. for automated runs via Cron)?

> thanks for the answer about the license, I don't know the details very well

No prob – until several years ago I didn't know either. But being one of the F-Droid maintainers (and running an F-Droid repo myself) it's part of my daily tasks :smile:

Please let me know when above changes have been checked in (I just looked, the code is not yet there) so I can cross-check again. I'd then volunteer to provide a `doc/getting_started.md` with instructions to get things running without digging in too deep (already did that with the setup of my PR, but will have to redo once your code is there). While on it, I could also "clean up" the current `README.md` a bit, e.g. moving specs to `doc/specifications.md` and then link to both documents from the Readme. Should save new users some confusion :wink:

**Comment by IzzySoft** [comment]
Hi @ya0211 – I just wonder were we currently stand? No pressure, just curiosity.

**Comment by ya0211** [comment]
- `util` needs a `push_rsync`, other parts are already done
https://github.com/ya0211/magisk-modules-repo-util/blob/20a5fa4f358d44e8cf8eff725dabb3d001e63a50/Sync/core/repo.py#L250-L251

- about MRepo View it at [Refactor MRepo](https://github.com/ya0211/MRepo/issues/4#issue-1533750683)

**Comment by IzzySoft** [comment]
Thanks for the update! So I'll need to wait until the client (MRepo) is ready.

Picking an item from there which applies here as well, even more maybe: Will "compatibility with established repos" be integrated, i.e. an option to (also) create a JSON in that format? There are other clients supporting user-defined repos, thus a "common format" would be great (as far as I can see the one other client I found, [MMRL](https://github.com/DerGoogler/MMRL), supports the format of those "established repos", and by default even includes one of them). So with this part established here, I could already fire up a repo as there is a client available for it.
